python/dgl/distributed/optim/pytorch/sparse_optim.py

```diff
@@ -35,16 +46,194 @@ def __init__(self, params, lr):
+                state = state.to(states[state_idx].dtype)
+                state = state.to(th.device("cpu"))
+                states[state_idx][idx] = state
+        self._defaults.update(local_state_dict["params"])
+        self.__dict__.update(local_state_dict["params"])
+
+    def save(self, f):
+        """Save the local state_dict to disk on per rank.
+
+        Saved dict contains three parts:
+
+        * 'ids': list of global ids of the nodes/edges selected locally
+        * 'params': hyper parameters of the optimizer
+        * 'states': partial optimizer states indexed by ```ids``` of entire states
+
+        NOTE: This needs to be called on all ranks.
+
+        Parameters
+        ----------
+        f : Union[str, PathLike]
+            The basic path of the file, client id
+            will be attached as full name.
+
+        See Also
+        --------
+        load
+        """
+        num_clients = dgl.distributed.get_num_client()
+        if num_clients > 1:
+            dgl.distributed.client_barrier()
+        client_id = dgl.distributed.get_rank()
+        f = f if isinstance(f, str) else str(f, "UTF-8")
+        f = f"{f}_{client_id}"
+        th.save(self.local_state_dict(), f)
+        if num_clients > 1:
+            dgl.distributed.client_barrier()
+
+    def load(self, f):
+        """Load the local state of the optimizer from the file on per rank.
+
+        NOTE: This needs to be called on all ranks.
+
+        Parameters
+        ----------
+        f : Union[str, PathLike]
+            The basic path of the file, client id
+            will be attached as full name.
+
+        See Also
+        --------
+        save
+        """
+        num_clients = dgl.distributed.get_num_client()
+        if num_clients > 1:
+            dgl.distributed.client_barrier()
+        client_id = dgl.distributed.get_rank()
+        f = f if isinstance(f, str) else str(f, "UTF-8")
+        f_attach_client_id = f"{f}_{client_id}"
+        # Don't throw error here to support device number scale-out
+        # after reloading, but make sure your hyper parameter is same
+        # as before because added local optimizers will not be filled
+        if not exists(f_attach_client_id):
+            warnings.warn(
+                f"File {f_attach_client_id} can't be found, load nothing."
+            )
+        else:
+            old_num_clients = self._load_state_from(f_attach_client_id)
+            # Device number scale-in
+            if num_clients < old_num_clients:
+                for rank in range(
+                    client_id + num_clients, old_num_clients, num_clients
+                ):
+                    self._load_state_from(f"{f}_{rank}")
+        if num_clients > 1:
+            dgl.distributed.client_barrier()
+
+    def _load_state_from(self, f):
+        local_state_dict = th.load(f)
+        num_clients = local_state_dict["params"]["num_clients"]
+        del local_state_dict["params"]["num_clients"]
```

Comment:
Why delete?

Reply:
It is used for ```load``` to determine scale-in but not actually for optimizer states.